state-network.md

```diff
@@ -0,0 +1,263 @@
+
+
+### Transport
+
+The lasted research suggests we need a network with long lived connections that allow streaming larger data payloads than are possible with single UDP packets.  This could be accomplished using libp2p or an implementation of [uTP](https://www.bittorrent.org/beps/bep_0029.html) over DiscV5.
+
+### Proofs
+
+As the main Ethereum network progresses, a proof will need to be produced that captures new, updated, and deleted trie data.  These proofs will contain some set of trie nodes.
+
+Each of the new and updated trie nodes will need to be gossiped to the nodes in the network that are interested in them.
+
+For deleted trie data, we will need some form of exclusion proof that allows nodes to schedule deletion of old trie data that is no longer part of the current state.
+
+### Messages
+
+The network uses the following message triple for transmission of proof data.
+
+#### AdvertiseProofs
+
+A message containing a list of identifiers, probably `content_ids` that the node has proofs available for.
+
+#### RetrieveProofs
+
+A message requesting one of the proofs advertised via an `AdvertiseProofs` message.
+
+#### Proofs
+
+The response message containing the proofs requested by a `RetrieveProofs` message.
+
+### DOS Mitigation
+
+We need mitigation against:
+
+- flooding the gossip network with lots of "valid" trie data that isn't new or updated.
+- bloated proofs which contain lots of *extra* trie nodes that aren't necessary (proofs should be minimal)
+- any form of amplification attack
+
+Requiring proofs to be "minimal" should be easy.
+
+Both the DOS and amplification attacks may be trivially solved by only having nodes gossip data that was of interest to them (new content that they didn't have but that falls into their area of interest).
+
+
+## DHT Network
+
+Our DHT will be an overlay network on the existing [Discovery V5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md) network.
+
+We use the same distance function as the core protocol.
+
+We use the same routing table structure as the core protocol.
+
+We use the same PING/PONG/FINDNODES/NODES messaging rules as the core protocol.
+
+### Content
+
+All content on the network has a key and a value.  The keys are referred to as `content_key`.  The value is referred to using the variable `content`. Each piece of content has a `content_id` which is derived from the `content_key` as `hash(content_key)`.
+
+> TODO: define `hash` function (probably sha256)
+
+The data stored in the network is referred to as "content".  The network houses three different types of content.
+
+- Account trie data
+- Contract storage trie data
+- Contract bytecode
+
+#### Account Trie Data
+
+```
+content_key  := content_type | trie_path | node_hash
+content_type := 0x01
```

Comment:
Inclusion of a `content_type` here seemed like good future proofing as well as eliminating the need to *infer* the content type from the *shape* of the `content_key`.

Supposing this network expands to become an archive node at some point... and the ethereum state undergoes one of the migrations to either binary trie or verkle trie... I would expect us to introduce new content types for the new trie data formats.  The same goes for code merklization....